OKTA-620013 Update images and text for OAuth & OIDC concept doc (#4268)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/concepts/oauth-openid/index.md

# OAuth 2.0 and OpenID Connect Overview

OAuth 2.0 and OpenID Connect (OIDC) are industry standard protocols for user authentication and authorization. Okta identity solutions are based on these standards.

---

**Learning outcomes**

* Learn how to implement flows based on OAuth 2.0 and OIDC using Okta.

> * Learn what flows and grant types are commonly used by different types of applications.

Reviewer: ```suggestion
* Learn which flows and grant types are commonly used by different types of applications.
```

> **Note**: To learn about the Okta authentication deployment models built on top of OAuth 2.0 and OIDC, see [Okta deployment models](/docs/concepts/redirect-vs-embedded/).

> ## OAuth 2.0 vs OpenID Connect

Reviewer: I think that 'vs' is latin and falls under terms that shouldn't be used.

> ## OAuth 2.0 vs OpenID Connect

Author: @timhowie-okta  versus counts as English these days. I'm going to leave this as is but add a period as it is an abbreviation, and that's how Microsoft does it. https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/a/argument-vs-parameter

OAuth 2.0 and OpenID Connect (OIDC) are complementary protocols, defining how users are authenticated by a server and then authorized to access resources:

* [OAuth 2.0](/docs/reference/api/oidc/) controls and delgates authorization to access a protected resource, like your web app, native app, or API service. It provides API security through scoped access tokens.

> * [OIDC](/docs/reference/api/oidc/) extends OAuth 2.0 with user authentication and Single Sign-On (SSO) functionality. It also enables you to retrieve and store authentication information about your end users. It also defines several OAuth 2.0 scopes to enable applications to access user profile information.

Reviewer: ```suggestion
* [OIDC](/docs/reference/api/oidc/) extends OAuth 2.0 with user authentication and Single Sign-On (SSO) functionality. It enables you to retrieve and store authentication information about your end users. It also defines several OAuth 2.0 scopes to enable applications to access user profile information.
```

Reviewer: Otherwise, you have "It also" as the start of two sentences in a row.

Okta recommends using one of its authentication deployment models for your application's authentication needs. They each abstract over the OAuth 2.0 and OIDC protocols, so you don't have to use them directly. To get started and to find sample apps, see [Sign users in](/docs/guides/sign-in-overview/).

> **Tip**: Use the [Authentication API](/docs/reference/api/authn/) if you require a custom app setup and workflow with direct access to your Okta org and app integrations. This API underpins both the Okta [redirect model](/docs/guides/sign-into-web-app-redirect/), [Embedded Sign-In Widget](/docs/guides/embedded-siw/), and [Auth JS](/docs/guides/auth-js/) SDKs.

### OAuth 2.0

OAuth 2.0 is a standard that apps use to provide client applications with access. If you would like to grant access to your application data in a secure way, then you want to use the OAuth 2.0 protocol.

The OAuth 2.0 spec has four important roles:

* **Client**: The application that wants to access some data.
* **Resource Server**: The API or application that stores the data the client wants to access.

> * **Resource Owner**: The owner of the data in the resource server. For example, you are the owner of your Facebook profile.

Reviewer: I might suggest the data is 'on' the server rather than 'in' the server, but I guess either works.
```suggestion
* **Resource Owner**: The owner of the data in the resource server. For example, you're the owner of your Facebook profile.
```

* **Authorization Server**: The server that manages access and issues access tokens. In this case Okta is the authorization server.

Other important terms:

> * **OAuth 2.0 grant**: The authorization given (or granted) to the client by the user. Examples of grants are **Authorization Code** and **Client Credentials**. Each OAuth grant has a corresponding flow. See [Choose an OAuth 2.0 flow](#choose-an-oauth-20-flow).

Author: @timhowie-okta Not sure. Have lowercased the previous list items.

## Choose an OAuth 2.0 flow

The OAuth flow that you use depends on your use case. The following sections recommend OAuth 2.0 flows based on:

* The type of application that you're building and the token types that the application requires
* The type of client that you're building

### What type of application are you building?

The following table shows you which OAuth 2.0 flow to use for the type of application that you're building.

| Type of application     | OAuth 2.0 flow / grant type                      | Access token?    | ID token?     |
| ----------------------- | ------------------------------------------------ | ---------------- | ------------- |
| Server-side (aka web), <br />Single-Page Application, <br />or Native | [Authorization Code with PKCE](#authorization-code-flow-with-pkce) or<br> [Interaction Code](#interaction-code-flow) (Identity Engine only).                | &#9989;          | &#9989;      |
| Trusted                 | [Interaction Code](#interaction-code-flow)            | &#9989;          | &#9989;      |
| Service                 | [Client Credentials](#client-credentials-flow)                  | &#9989;          | &#10060;     |

> > **Note**: The OAuth 2.0 [SAML 2.0 Assertion flow](#saml-20-assertion-flow) is also available. This flow is intended for client applications that want to use an existing trust relationship without a direct user approval step at the authorization server. It supports access and ID tokens.

Reviewer: I think the "There's also a SAML 2.0 assertion flow." seems like a more natural phrasing.

Author: Sure - I'll buy that.

#### Does your client use the redirect or embedded authentication model?

If your client application is a SPA or a native application that redirects its authentication requests to an Okta-hosted sign-in page ([the redirect model](/docs/concepts/redirect-vs-embedded/#redirect-authentication)), use the [Authorization Code flow with PKCE](#authorization-code-flow-with-pkce). If your application hosts the authentication flow itself ([the embedded model](/docs/concepts/redirect-vs-embedded/#embedded-authentication)), use the [Interaction Code flow](#interaction-code-flow).

#### Does the client have an end user?

If your client application runs on a server with no direct end user, then it can be trusted with its own credentials and use them responsibly. If your client application is only doing machine-to-machine interaction, then you should use the [Client Credentials flow](#client-credentials-flow).

#### Is your app high-trust?

> Applications are **high-trust** if you own both them and the resource they access. Because you own them, you can trust them to handle your end users' usernames and passwords. In this case, and _only if other flows aren't viable_, you may use the [Resource Owner Password flow](#resource-owner-password-flow). However, it isn't possible to use this flow with multifactor authentication, so you will likely want to look at alternatives such as the [Authorization Code](/docs/guides/implement-grant-type/authcode/main/) or [Interaction Code flow](#interaction-code-flow).

Reviewer: ```suggestion
Applications are **high-trust** if you own both the app and the resource that it accesses. Because you own both, you can trust them to handle your end users' usernames and passwords. In this case, and _only if other flows aren't viable_, you can use the [Resource Owner Password flow](#resource-owner-password-flow). However, it isn't possible to use this flow with multifactor authentication, so you'll want to look at alternatives such as the [Authorization Code](/docs/guides/implement-grant-type/authcode/main/) or [Interaction Code flow](#interaction-code-flow).
```

If your app isn't high-trust, or if you want to take advantage of multifactor authentication, you should use the [Authorization Code](/docs/guides/implement-grant-type/authcode/main/) flow.

### Authorization Code flow with PKCE

Proof Key for Code Exchange (PKCE) was originally designed as an extension to protect the Authorization Code flow in mobile apps. However, its ability to prevent authorization code injection and keep the flow secure makes it optimal for every type of OAuth client. Okta recommends you use the Authorization Code flow with PKCE for your OAuth client if possible.

The flow requires your application to generate a cryptographically random string called a **code verifier**. The code verifier is then hashed to create the **code challenge**, and this challenge is passed along with the request for the authorization code.

> After the application receives the authorization code, it sends the authorization code and the code verifier in a request for an access token. The authorization server recomputes the challenge from the verifier using the previously agreed-upon hash algorithm and then compares that. If the two code challenges and verifier match, the authorization server knows that the same client sent both requests.

Reviewer 2: In the second sentence "and then compares that" isn't clear what is being compared and what "that" is. Can you please rephrase?

```suggestion
After the application receives the authorization code, it sends the authorization code and the code verifier in a request for an access token. The authorization server recomputes the challenge from the verifier using the previously agreed-upon hash algorithm and then compares that. If the two code challenges and verifier match, the authorization server knows that the same client sent both requests.
```

Author: Have changed it to

> The flow requires your application to generate a cryptographically random string called a **code verifier**. The code verifier is then hashed to create the **code challenge**, and this challenge is passed along with the request for the authorization code. The authorization server responds with an authorization code and associates the code challenge with the authorization code.
> 
> After the application receives the authorization code, it sends the authorization code and the code verifier in a request for an access token. The authorization server recomputes the challenge from the verifier using the previously agreed-upon hash algorithm and then compares the challenge with the one it associated with the authorization code in the previous step. If the two code challenges and verifier match, the authorization server knows that the same client sent both requests.

participant "Resource Server (Your App)" as app

user -> client: Start auth with user info
client -> client: Generate PKCE code verifier & challenge
client -> okta: Authorization request w/ code_challenge, client ID, scopes, and user info
okta -> okta: Remediation required
okta -> client: Send interaction_handle in response (for required interaction)
client <-> okta: Remediation steps w/ interaction_handle
user <-> client: Remediation
note right: Possible multiple remediation steps required
client -> okta: Complete remediation steps w/ interaction_handle
okta -> client: Send interaction_code in response
client -> okta: Send interaction_code, client ID, code_verifier to /token
okta -> okta: Evaluates PKCE code verifier
okta -> client: Access token (and optionally refresh token)
client -> app: Request with access token
app -> client: Response
@enduml

 -->

### Resource Owner Password flow

The Resource Owner Password flow is intended for use cases where:

* You control both the client application and the resource that it's interacting with.
* The client can store a client secret and can be trusted with the resource owner's credentials.

> * You do not need your users to use multifactor authentication.

Reviewer: ```suggestion
* You don't need your users to use multifactor authentication.
```

# OAuth 2.0 and OpenID Connect Overview

OAuth 2.0 and OpenID Connect (OIDC) are industry standard protocols for user authentication and authorization. Okta identity solutions are based on these standards.

---

**Learning outcomes**

* Learn how to implement flows based on OAuth 2.0 and OIDC using Okta.
* Learn what flows and grant types are commonly used by different types of applications.

> **Note**: To learn about the Okta authentication deployment models built on top of OAuth 2.0 and OIDC, see [Okta deployment models](/docs/concepts/redirect-vs-embedded/).

## OAuth 2.0 vs OpenID Connect

> OAuth 2.0 and OpenID Connect (OIDC) are complementary protocols, defining how users are first authenticated by a server and then authorized to access resources:

Reviewer 3: I wouldn't use "first" here because it isn't really a sequence of events. This makes it sound like the user is authenticated at Okta using OIDC, and then later uses OAuth to get authorized to access resources. In reality, authentication at Okta is using Okta's own internal mechanisms like passwords, email magic link, passkeys, etc. Users then use OIDC and/or OAuth to access applications.

Author: Fair enough. I'll simply remove "first".

> * [OAuth 2.0](/docs/reference/api/oidc/) controls and delgates authorization to access a protected resource, like your web app, platform-specific app, or API service. It provides API security through scoped access tokens.

Reviewer 3: again I'm not sure where "platform-specific" came from

> * [OIDC](/docs/reference/api/oidc/) extends OAuth 2.0 with user authentication and Single Sign-On (SSO) functionality. It also enables you to retrieve and store authentication information about your end users. It's also more opinionated than plain OAuth 2.0, for example in its scope definitions.

Reviewer 3: ```suggestion
* [OIDC](/docs/reference/api/oidc/) extends OAuth 2.0 with user authentication and Single Sign-On (SSO) functionality. It also enables you to retrieve and store authentication information about your end users. It also defines several OAuth 2.0 scopes to enable applications to access user profile information.
```

Author: Thank you for the wording. Updated.

Okta recommends using one of its authentication deployment models for your application's authentication needs. They each abstract over the OAuth 2.0 and OIDC protocols, so you don’t have to use them directly. To get started and to find sample apps, see [Sign users in](/docs/guides/sign-in-overview/).

> **Tip**: Use the [Authentication API](/docs/reference/api/authn/) if you require a custom app setup and workflow with direct access to your Okta org and app integrations. This API underpins both the Okta [redirect model](/docs/guides/sign-into-web-app-redirect/), [Embedded Sign-In Widget](/docs/guides/embedded-siw/), and [Auth JS](/docs/guides/auth-js/) SDKs.

### OAuth 2.0

OAuth 2.0 is a standard that apps use to provide client applications with access. If you would like to grant access to your application data in a secure way, then you want to use the OAuth 2.0 protocol.

The OAuth 2.0 spec has four important roles:

* **Client**: The application that wants to access some data.
* **Resource Owner**: The owner of the data in the resource server. For example, you are the owner of your Facebook profile.

> * **Resource Server**: The API or application that stores the data the client wants to access.

Reviewer 3: Since you define resource owner relative to the resource server, makes more sense to move resource server before resource owner.

> * **Authorization Server**: The main engine of OAuth. In this case Okta is the authorization server.

Reviewer 3: ```suggestion
* **Authorization Server**: The server that manages access and issues access tokens. In this case Okta is the authorization server.
```

Other important terms:

* **OAuth 2.0 grant**: The authorization given (or granted) to the client by the user. Examples of grants are **Authorization Code** and **Client Credentials**. Each OAuth grant has a corresponding flow. See [Choose an OAuth 2.0 flow](#choose-an-oauth-20-flow).
* **access token**: The token issued by the authorization server (Okta) in exchange for the grant.
* **refresh token**: An optional token that is exchanged for a new access token if the access token has expired.

> **Note:** For more information on hard-coded and configurable token lifetimes, see [Token lifetime](/docs/reference/api/oidc/#token-lifetime).

> The usual OAuth 2.0 grant flow looks like this:

Reviewer 3: ```suggestion
The usual OAuth 2.0 authorization code flow looks like this:
```

Author: updated

* **relying party**: The client application that requests the ID token from Okta.
* **ID token**: The token issued by the OpenID provider and contains information about the end user in the form of claims.
* **claim**: The claim is a piece of information about the end user.

The high-level flow looks the same for both OpenID Connect and regular OAuth 2.0 flows. The primary difference is that an OpenID Connect flow results in an ID token, in addition to any access or refresh tokens.

## Choose an OAuth 2.0 flow

The OAuth flow that you use depends on your use case. The following sections recommend OAuth 2.0 flows based on:

* The type of application that you're building and the token types that the application requires
* The type of client that you're building

### What type of application are you building?

The following table shows you which OAuth 2.0 flow to use for the type of application that you’re building.

| Type of application     | OAuth 2.0 flow / grant type                      | Access token?    | ID token?     |
| ----------------------- | ------------------------------------------------ | ---------------- | ------------- |

> | Server-side (aka web), <br />Single-Page Application, <br />or Platform-specific | [Authorization Code with PKCE](#authorization-code-flow-with-pkce) or<br> [Interaction Code](#interaction-code-flow) (Identity Engine only).                | &#9989;          | &#9989;      |

Reviewer 3: Where did "platform-specific" come from? That isn't a term we usually use in OAuth.

Author: @aaronpk We're actively trying to avoid using the word "Native" for inclusivity reasons on the docs. "Platform-specific" is the standard alternative adjective as it covers mobile and desktop apps.

Author: Agreed upon Mobile or Desktop instead.

> | Trusted                 | [Resource Owner Password](#resource-owner-password-flow)            | &#9989;          | &#9989;      |

Reviewer 3: Can we use Interaction Code here for "Trusted" apps instead of password? We don't really recommend the password flow at all anymore.

Author: Of course

| Service                 | [Client Credentials](#client-credentials-flow)                  | &#9989;          | &#10060;     |

> **Note**: The OAuth 2.0 [SAML 2.0 Assertion flow](#saml-20-assertion-flow) is also available. This flow is intended for client applications that want to use an existing trust relationship without a direct user approval step at the authorization server. It supports access and ID tokens.

### What kind of client are you building?

The type of OAuth 2.0 flow depends on what kind of client that you’re building. This flowchart can quickly help you decide which flow to use.

<div class="full border">

   ![Decision tree for choosing the correct OAuth 2.0 flow based on the type of client being built](/img/authorization/oauth_grant_flowchart_new.png)

   <!--
      Source image: https://www.figma.com/file/YH5Zhzp66kGCglrXQUag2E/%F0%9F%93%8A-Updated-Diagrams-for-Dev-Docs?type=design&node-id=4162%3A41793&mode=design&t=uYvvxmVdQRo6BVjR-1 oauth_grant_flowchart_new
   -->

</div>

#### Is your client public?

> Single-Page Applications (SPAs), mobile, and platform-specific applications are **public** applications where end users can view and possibly modify their code. Any secrets in their code are open to malicious users. By comparison, server-side (web) applications are **confidential** or **private** applications. Private application clients can use client-side authentication methods such as client secrets.

Reviewer 3: ```suggestion
Single-Page Applications (SPAs), mobile, and platform-specific applications are **public** applications where end users can view and possibly modify the source code code of the application. Any secrets in the code are exposed to malicious users. By comparison, server-side (web) applications are **confidential** or **private** applications. Confidential clients can use client-side authentication methods such as client secrets and private keys.
```

Author: Updated

#### Does your client use the redirect or embedded authentication model?

If your client application is a SPA or a platform-specfic application that redirects its authentication requests to an Okta-hosted sign-in page ([the redirect model](/docs/concepts/redirect-vs-embedded/#redirect-authentication)), use the [Authorization Code flow with PKCE](#authorization-code-flow-with-pkce). If your application hosts the authentication flow itself ([the embedded model](/docs/concepts/redirect-vs-embedded/#embedded-authentication)), use the [Interaction Code flow](#interaction-code-flow).

#### Does the client have an end user?

> If your client application runs on a server with no direct end user, then it can be trusted to handle credentials and use them responsibly. If your client application is only doing machine-to-machine interaction, then you should use the [Client Credentials flow](#client-credentials-flow).

Reviewer 3: ```suggestion
If your client application runs on a server with no direct end user, then it can be trusted with its own credentials and use them responsibly. If your client application is only doing machine-to-machine interaction, then you should use the [Client Credentials flow](#client-credentials-flow).
```

#### Is your app high-trust?

> Applications are **high-trust** if you own both them and the resource they access. Because you own them, you can trust them to handle your end users' usernames and passwords. In this case, and _only if other flows aren't viable_, you may use the [Resource Owner Password flow](#resource-owner-password-flow).

Reviewer 3: ```suggestion
Applications are **high-trust** if you own both them and the resource they access. Because you own them, you can trust them to handle your end users' usernames and passwords. In this case, and _only if other flows aren't viable_, you may use the [Resource Owner Password flow](#resource-owner-password-flow). However, it isn't possible to use this flow with multifactor authentication, so you will likely want to look at alternatives such as the [Authorization Code](/docs/guides/implement-grant-type/authcode/main/) or [Interaction Code flow](#interaction-code-flow).
```

Author: Updated

> If your app isn’t high-trust, you should use the [Authorization Code](/docs/guides/implement-grant-type/authcode/main/) flow.

Reviewer 3: ```suggestion
If your app isn’t high-trust, or if you want to take advantage of multifactor authentication, you should use the [Authorization Code](/docs/guides/implement-grant-type/authcode/main/) flow.
```

### Authorization Code flow with PKCE

Proof Key for Code Exchange (PKCE) was originally designed as an extension to protect the Authorization Code flow in mobile apps. However, its ability to prevent authorization code injection and keep the flow secure makes it optimal for every type of OAuth client. Okta recommends you use the Authorization Code flow with PKCE for your OAuth client if possible.

> The flow requires your application to generate a cryptographically random key called a **code verifier**. A **code challenge** is then created from the verifier, and this challenge is passed along with the request for the authorization code.

Reviewer 3: ```suggestion
The flow requires your application to generate a cryptographically random string called a **code verifier**. The code verifier is then hashed to create the **code challenge**, and this challenge is passed along with the request for the authorization code.
```

> After the code is received, it’s sent with the code verifier in a request for an access token. The authorization server recomputes the challenge from the verifier using an agreed-upon hash algorithm and then compares that. If the two code challenges and verifier match, the authorization server knows that the same client sent both requests.

Reviewer 3: ```suggestion
After the application receives the authorization code, it sends the authorization code along with the code verifier in a request for an access token. The authorization server recomputes the challenge from the verifier using the previously agreed-upon hash algorithm and then compares that. If the two code challenges match, the authorization server knows that the same client sent both requests.
```

Author: updated

> **Note:** For implementing refresh tokens with SPAs and other browser-based apps, see [Refresh access tokens](/docs/guides/refresh-tokens/main/).

> If a rogue app intercepts the authorization code, it still doesn't have access to the code challenge or verifier, since they’re both sent over HTTPS.

Reviewer 3: ```suggestion
```

   -->
</div>

<!-- Source for image. Generated using http://www.plantuml.com/plantuml/uml/

@startuml

skinparam monochrome true

actor "Resource Owner (User)" as user
participant "Client" as client
participant "Authorization Server (Okta)" as okta
participant "Resource Server (Your App)" as app

client -> client: Generate PKCE code verifier & challenge
client -> okta: Authorization Code Request + code_challenge to /authorize
okta -> user: 302 redirect to authentication prompt
user -> okta: Authentication & consent
okta -> client: Authorization Code Response
client -> okta: Send authorization code + code_verifier to /token
okta -> okta: Evaluates PKCE code
okta -> client: Access token (and optionally Refresh Token)
client -> app: Request with access token
app -> client: Response

-->

For information on how to set up your application to use this flow, see [Implement the Authorization Code flow with PKCE](/docs/guides/implement-grant-type/authcodepkce/main/).

### Interaction Code flow

> The Interaction Code flow extends the OAuth 2.0 and OIDC standards. It requires clients to pass a client ID and PKCE to Okta to keep the flow secure. The user can start the request with minimal information, relying on the client to facilitate the interactions with Okta to authenticate the user. See [Interaction Code grant type](/docs/concepts/interaction-code/).

Reviewer 3: ```suggestion
The Interaction Code flow extends the OAuth 2.0 and OIDC standards. It requires clients to pass a client ID and PKCE parameters to Okta to keep the flow secure. The user can start the request with minimal information, relying on the client to facilitate the interactions with Okta to authenticate the user. See [Interaction Code grant type](/docs/concepts/interaction-code/).
```

Author: updated

   <!--
      Source image: https://www.figma.com/file/YH5Zhzp66kGCglrXQUag2E/%F0%9F%93%8A-Updated-Diagrams-for-Dev-Docs?type=design&node-id=4133%3A43899&mode=design&t=Me7qqw8odOmrLh6K-1
      oauth-interaction-code-grant-flow
   -->
</div>

<!--
See http://www.plantuml.com/plantuml/uml/

@startuml
skinparam monochrome true
actor "Resource Owner (User)" as user
participant "Client" as client
participant "Authorization Server (Okta)" as okta
participant "Resource Server (Your App)" as app

user -> client: Start auth with user info
client -> client: Generate PKCE code verifier & challenge
client -> okta: Authorization request w/ code_challenge, client ID, scopes, and user info
okta -> okta: Remediation required
okta -> client: Send interaction_handle in response (for required interaction)
client <-> okta: Remediation steps w/ interaction_handle
user <-> client: Remediation
note right: Possible multiple remediation steps required
client -> okta: Complete remediation steps w/ interaction_handle
okta -> client: Send interaction_code in response
client -> okta: Send interaction_code, client ID, code_verifier to /token

> okta -> okta: Evaluates PKCE code

Reviewer 3: ```suggestion
okta -> okta: Evaluates PKCE code_verifier
```

okta -> client: Access token (and optionally refresh token)
client -> app: Request with access token
app -> client: Response
@enduml

 -->

### Resource Owner Password flow

The Resource Owner Password flow is intended for use cases where:

* You control both the client application and the resource that it’s interacting with.

> * The client can store a client secret and can be trusted with the resource owner's credentials.

Reviewer 3: ```suggestion
* The client can store a client secret and can be trusted with the resource owner's credentials.
* You do not need your users to use multifactor authentication.
```

Author: updated

> It is most commonly found in clients made for online services, like the Facebook client applications that interact with the Facebook service. It doesn't require redirects like the Authorization Code or Implicit flows, and involves a single authenticated call to the `/token` endpoint.

Author: updated

### Implicit flow

 > **Note:** The Implicit flow is a legacy flow used only for SPAs that can't support PKCE.

> The Implicit flow was intended for browser-based applications that didn't support Cross-Origin Resource Sharing (CORS) and lacked modern cryptography APIs, and that couldn't protect a client secret. In this flow, the client doesn't make a request to the `/token` endpoint, but instead receives the access token in the redirect from the `/authorize` endpoint. The client must be able to interact with the resource owner's user agent and to receive incoming requests (through redirection) from the authorization server.

Reviewer 2: ```suggestion
The Implicit flow was intended for browser-based applications that didn't support Cross-Origin Resource Sharing (CORS), lacked modern cryptography APIs, and that couldn't protect a client secret. In this flow, the client doesn't make a request to the `/token` endpoint, but instead receives the access token in the redirect from the `/authorize` endpoint. The client must be able to interact with the resource owner's user agent and to receive incoming requests (through redirection) from the authorization server.
```

 > **Note:** The Implicit flow is a legacy flow used only for SPAs that can’t support PKCE.

>  The Implicit flow is intended for applications where the confidentiality of the client secret can't be guaranteed. In this flow, the client doesn't make a request to the `/token` endpoint, but instead receives the access token directly from the `/authorize` endpoint. The client must be able to interact with the resource owner's user agent and to receive incoming requests (through redirection) from the authorization server.

Reviewer 3: ```suggestion
The Implicit flow was intended for browser-based applications that didn't support Cross-Origin Resource Sharing (CORS), and that couldn't protect a client secret. In this flow, the client doesn't make a request to the `/token` endpoint, but instead receives the access token in the redirect from the `/authorize` endpoint. The client must be able to interact with the resource owner's user agent and to receive incoming requests (through redirection) from the authorization server.
```

Author: updated

> > **Note:** Because it's intended for less-trusted clients, the Implicit flow doesn't support refresh tokens.

Reviewer 3: ```suggestion
> **Note:** Because it was always intended for less-trusted clients, the Implicit flow doesn't support refresh tokens.
```
